go/types/opcache.go

```diff
@@ -91,37 +125,100 @@ func encToSlice(v Value, initBuf []byte, vw ValueWriter) []byte {
+		entry.key = DecodeFromBytes(ldbValue[uint32Size:uint32Size+keyBytesLen], i.vr, staticTypeCache)
+		entry.value = DecodeFromBytes(ldbValue[uint32Size+keyBytesLen:], i.vr, staticTypeCache)
 	}
 
-	entry.value = DecodeFromBytes(data[dataOffset:], i.vr, staticTypeCache)
 	return entry
 }
 
-func (i *opCacheIterator) Release() {
+// Insert can be called from any goroutine
+func (opc *ldbOpCache) SetInsert(val Value) {
+
+	switch val.Type().Kind() {
+	default:
+		ldbKey := ldbKeyFromValueHash(val, opc.colId)
+		valArray := [initialBufferSize]byte{}
+		encodedVal := encToSlice(val, valArray[:], opc.vrw)
+		err := opc.store.ldb.Put(ldbKey, encodedVal, nil)
+		d.Chk.NoError(err)
+
+	case BoolKind, NumberKind, StringKind:
+		ldbKey := ldbKeyFromValue(val, opc.colId, opc.vrw)
+		// Since the ldbKey contains the val, there's no reason to store anything in the ldbValue
+		err := opc.store.ldb.Put(ldbKey, nil, nil)
+		d.Chk.NoError(err)
+	}
+}
+
+func (i *ldbOpCacheIterator) SetOp() sequenceItem {
+	ldbKey := i.iter.Key()
+
+	switch NomsKind(ldbKey[uint32Size]) {
+	case BoolKind, NumberKind, StringKind:
+		return DecodeFromBytes(ldbKey[uint32Size:], i.vr, staticTypeCache)
+	default:
+		data := i.iter.Value()
+		return DecodeFromBytes(data, i.vr, staticTypeCache)
+	}
+}
+
+func (i *ldbOpCacheIterator) Release() {
 	i.iter.Release()
 }
+
+// writeLdbKeyHeaderBytes writes the first 4 or 5 bytes into the ldbKey. The first 4 bytes in every
+// ldbKey are the colId. This identifies all the keys for a particular opCache and allows this opStore
+// to cache values for multiple collections. The optional 5th byte is the NomsKind of the value. In
+// cases where we're encoding the hash of an object we need to write the nomsKind manually because the
+// hash doesn't contain it. In cases were we are encoding a primitive value into the key, the first byte
+// of the value is the nomsKind so there is no reason to write it again.
+func writeLdbKeyHeaderBytes(ldbKey []byte, colId uint32, v Value) []byte {
+	binary.LittleEndian.PutUint32(ldbKey, colId)
+	length := uint32Size
+	if v != nil {
+		ldbKey[length] = byte(v.Type().Kind())
+		length++
+	}
+	return ldbKey[0:length]
+}
+
+// ldbKeys for non-primitive Nom Values (e.g. blobs, structs, lists, maps, etc) use a serialization of the values hash:
+// colId(4 bytes) + nomsKind(val)(1 byte) + val.Hash()(20 bytes).
+
+func ldbKeyFromValueHash(val Value, colId uint32) []byte {
+	ldbKeyArray := [uint32Size + 1 + hash.ByteLen]byte{}
+	ldbKey := writeLdbKeyHeaderBytes(ldbKeyArray[:], colId, val)
+	return append(ldbKey, val.Hash().DigestSlice()...)
+}
+
+// ldbKeys for primitive Noms Values (e.g. bool, number, & string) consist of a byte string that encodes:
+// colId(4 bytes) + serialized value(n bytes)
+// Note: the first byte of the serialized value is the NomsKind.
+func ldbKeyFromValue(val Value, colId uint32, vrw ValueReadWriter) []byte {
+	valArray := [initialBufferSize]byte{}
+	ldbKeyArray := [initialBufferSize]byte{}
```

Comment:
You know that this is going to be at most 5 bytes, right? So you can just allocate that much space here.